File: tools/visual_qa_dashboard.py
import json
import os
import sys
from pathlib import Path

from playwright.sync_api import sync_playwright


EDGE_PATH = r"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"


def main() -> int:
    output_dir = Path("tmp")
    output_dir.mkdir(parents=True, exist_ok=True)
    dashboard_url = os.environ.get("DASHBOARD_URL", "http://localhost:8501")
    results = []
    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=True, executable_path=EDGE_PATH)
        for name, width, height in (("desktop", 1440, 900), ("mobile", 390, 844)):
            page = browser.new_page(viewport={"width": width, "height": height})
            page.goto(dashboard_url, wait_until="networkidle", timeout=30000)
            page.get_by_role("heading", name="A股市场复盘 Dashboard").wait_for(timeout=20000)
            page.get_by_text("正式真实数据", exact=False).wait_for(timeout=20000)
            page.wait_for_timeout(1000)
            for _ in range(8):
                page.mouse.wheel(0, height)
                page.wait_for_timeout(250)
            page.mouse.wheel(0, -height * 8)
            page.wait_for_timeout(500)
            body = page.locator("body").inner_text()
            diagnostics = page.evaluate(
                """() => ({
                    overflow: document.documentElement.scrollWidth - document.documentElement.clientWidth,
                    svgCount: document.querySelectorAll("svg").length,
                    dataframeCount: document.querySelectorAll('[data-testid="stDataFrame"]').length
                })"""
            )
            result = {
                "name": name,
                "title": page.title(),
                "hasDashboard": "A股市场复盘 Dashboard" in body,
                "hasMarket": "今日市场" in body,
                "hasThemes": "今日主线 TOP5" in body,
                "hasQuality": "正式真实数据" in body,
                "hasException": "Traceback" in body or "Exception" in body,
                **diagnostics,
            }
            results.append(result)
            page.screenshot(path=str(output_dir / f"dashboard-{name}.png"), full_page=True)
            page.close()
        browser.close()
    print(json.dumps(results, ensure_ascii=False, indent=2))
    if any(
        not item["hasDashboard"]
        or not item["hasMarket"]
        or not item["hasThemes"]
        or not item["hasQuality"]
        or item["hasException"]
        or item["overflow"] > 2
        for item in results
    ):
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/visual_qa_dashboard.js
const { chromium } = require("playwright");
const fs = require("fs");
const path = require("path");

const edgePath = "C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe";
const outputDir = path.resolve("tmp");
const dashboardUrl = process.env.DASHBOARD_URL || "http://localhost:8501";
fs.mkdirSync(outputDir, { recursive: true });

(async () => {
  const browser = await chromium.launch({ headless: true, executablePath: edgePath });
  const viewports = [
    ["desktop", 1440, 900],
    ["mobile", 390, 844],
  ];
  const results = [];
  for (const [name, width, height] of viewports) {
    const page = await browser.newPage({ viewport: { width, height } });
    await page.goto(dashboardUrl, { waitUntil: "networkidle", timeout: 30000 });
    await page.getByRole("heading", { name: "A股市场复盘 Dashboard" }).waitFor({ timeout: 20000 });
    await page.getByText("正式真实数据", { exact: false }).waitFor({ timeout: 20000 });
    await page.waitForTimeout(1000);
    const body = await page.locator("body").innerText();
    const diagnostics = await page.evaluate(() => ({
      overflow: document.documentElement.scrollWidth - document.documentElement.clientWidth,
      svgCount: document.querySelectorAll("svg").length,
      dataframeCount: document.querySelectorAll('[data-testid="stDataFrame"]').length,
    }));
    const result = {
      name,
      title: await page.title(),
      hasDashboard: body.includes("A股市场复盘 Dashboard"),
      hasMarket: body.includes("今日市场"),
      hasThemes: body.includes("今日主线 TOP5"),
      hasQuality: body.includes("正式真实数据"),
      hasException: body.includes("Traceback") || body.includes("Exception"),
      ...diagnostics,
    };
    results.push(result);
    await page.screenshot({ path: path.join(outputDir, `dashboard-${name}.png`), fullPage: true });
    await page.close();
  }
  await browser.close();
  console.log(JSON.stringify(results, null, 2));
  if (results.some((item) => !item.hasDashboard || !item.hasMarket || !item.hasThemes || !item.hasQuality || item.hasException || item.overflow > 2)) {
    process.exit(1);
  }
})();
